File: src/plugins/botbuilder/commands/billing.ts
import type { CommandDef } from '../../../core/plugin/types.js'
import type { BotbuilderStorage } from '../storage.js'
import { CREDIT_PACKS, HOSTING_PLANS } from '../config.js'
import type { CreditPackKey, HostingPlanKey } from '../config.js'

function yoomoneyLink(wallet: string, amount: number, label: string, comment: string): string {
  const params = new URLSearchParams({
    receiver: wallet,
    'quickpay-form': 'button',
    paymentType: 'AC',
    sum: String(amount),
    label,
    comment,
  })
  return `https://yoomoney.ru/quickpay/confirm.xml?${params}`
}

export function balanceCommand(storage: BotbuilderStorage): CommandDef {
  return {
    name: 'balance',
    description: 'Показать баланс кредитов',
    category: 'plugin',
    async handler(args) {
      const user = await storage.getOrCreateUser(args.channelId, args.userId)
      const text = `💰 Твой баланс: <b>${user.credits} кредитов</b>\n\nКредиты используются для создания и улучшения ботов.`
      return { type: 'adaptive', fallback: `💰 Баланс: ${user.credits} кредитов`, variants: { telegram: { text, parse_mode: 'HTML' } } }
    },
  }
}

export function buyCommand(yoomoneyWallet: string): CommandDef {
  return {
    name: 'buy',
    description: 'Купить пакет кредитов',
    usage: '[start|standard|pro]',
    category: 'plugin',
    async handler(args) {
      const packKey = args.raw.trim().toLowerCase() as CreditPackKey
      const pack = CREDIT_PACKS[packKey]

      if (!pack) {
        const lines = Object.entries(CREDIT_PACKS).map(([k, p]) =>
          `• <b>${p.label}</b> — ${p.credits} кред. за ${p.price_rub} ₽ / ${p.stars} ⭐\n  /buy ${k}`,
        )
        const text = `💳 <b>Пакеты кредитов:</b>\n\n${lines.join('\n\n')}\n\n🎁 Новым пользователям — 10 кредитов бесплатно`
        return { type: 'adaptive', fallback: text, variants: { telegram: { text, parse_mode: 'HTML' } } }
      }

      const label = `credits_${packKey}_${args.userId}_${Date.now()}`
      const link  = yoomoneyLink(yoomoneyWallet, pack.price_rub, label, `Кредиты ${pack.label}`)

      const text = (
        `💳 <b>Пакет "${pack.label}"</b> — ${pack.credits} кредитов\n\n` +
        `Способ оплаты:\n` +
        `⭐ <b>Telegram Stars:</b> ${pack.stars} звёзд (в разработке)\n` +
        `💳 <b>ЮMoney:</b> ${pack.price_rub} ₽ → <a href="${link}">Оплатить</a>`
      )
      return { type: 'adaptive', fallback: text, variants: { telegram: { text, parse_mode: 'HTML', disable_web_page_preview: true } } }
    },
  }
}

export function hostingCommand(storage: BotbuilderStorage, yoomoneyWallet: string): CommandDef {
  return {
    name: 'hosting',
    description: 'Продлить хостинг бота',
    usage: '[bot-id] [week|month|year]',
    category: 'plugin',
    async handler(args) {
      const parts   = args.raw.trim().split(/\s+/)
      const botId   = parts[0]
      const planKey = (parts[1] ?? '') as HostingPlanKey

      if (!botId) {
        const bots = await storage.getUserBots(args.userId)
        if (!bots.length) return { type: 'text', text: 'У тебя нет ботов. Создай: /create-bot' }

        const lines = bots.map((b) => {
          const until  = b.hostingUntil ? b.hostingUntil.slice(0, 10) : '—'
          const status = b.isActive ? '✅' : '⛔'
          return `${status} <b>${b.name}</b> — до ${until}\n  /hosting ${b.id} week | month | year`
        })
        const text = `🖥 <b>Твои боты:</b>\n\n${lines.join('\n\n')}`
        return { type: 'adaptive', fallback: text, variants: { telegram: { text, parse_mode: 'HTML' } } }
      }

      const plan = HOSTING_PLANS[planKey]
      if (!plan) {
        const opts = Object.entries(HOSTING_PLANS)
          .map(([k, p]) => `• ${p.label} — ${p.price_rub} ₽ → /hosting ${botId} ${k}`)
          .join('\n')
        const text = `Выбери период хостинга:\n\n${opts}`
        return { type: 'text', text }
      }

      const label = `hosting_${planKey}_${botId}_${Date.now()}`
      const link  = yoomoneyLink(yoomoneyWallet, plan.price_rub, label, `Хостинг ${plan.label}`)

      const text = (
        `🖥 <b>Хостинг "${plan.label}"</b>\n\n` +
        `⭐ Telegram Stars: ${plan.stars} звёзд (в разработке)\n` +
        `💳 ЮMoney: ${plan.price_rub} ₽ → <a href="${link}">Оплатить</a>\n\n` +
        `После оплаты напиши: /hosting-confirm ${botId} ${planKey}`
      )
      return { type: 'adaptive', fallback: text, variants: { telegram: { text, parse_mode: 'HTML', disable_web_page_preview: true } } }
    },
  }
}

export function hostingConfirmCommand(storage: BotbuilderStorage): CommandDef {
  return {
    name: 'hosting-confirm',
    description: 'Подтвердить оплату хостинга (после оплаты ЮMoney)',
    usage: '<bot-id> <week|month|year>',
    category: 'plugin',
    async handler(args) {
      const [botId, planKey] = args.raw.trim().split(/\s+/)
      const plan = HOSTING_PLANS[planKey as HostingPlanKey]
      if (!botId || !plan) return { type: 'error', message: 'Использование: /hosting-confirm <bot-id> <week|month|year>' }

      const newUntil = await storage.extendHosting(args.userId, botId, plan.days)
      const text = `✅ Хостинг продлён до <b>${newUntil.slice(0, 10)}</b>`
      return { type: 'adaptive', fallback: text, variants: { telegram: { text, parse_mode: 'HTML' } } }
    },
  }
}

export function myBotsCommand(storage: BotbuilderStorage): CommandDef {
  return {
    name: 'my-bots',
    description: 'Список всех созданных ботов',
    category: 'plugin',
    async handler(args) {
      const bots = await storage.getUserBots(args.userId)
      if (!bots.length) return { type: 'text', text: 'У тебя пока нет ботов. Создай: /create-bot' }

      const lines = bots.map((b) => {
        const until  = b.hostingUntil ? b.hostingUntil.slice(0, 10) : '—'
        const status = b.isActive ? '✅ Активен' : '⛔ Остановлен'
        return `🤖 <b>${b.name}</b>\n   ${status} | до ${until}\n   <a href="${b.dashboardUrl}">Dashboard</a>`
      })
      const text = `📋 <b>Твои боты:</b>\n\n${lines.join('\n\n')}`
      return { type: 'adaptive', fallback: text, variants: { telegram: { text, parse_mode: 'HTML', disable_web_page_preview: true } } }
    },
  }
}

File: bot/handlers/billing.py
from aiogram import Router, F, Bot
from aiogram.types import (
    Message, CallbackQuery, InlineKeyboardMarkup, InlineKeyboardButton,
    PreCheckoutQuery, ReplyKeyboardRemove
)
from aiogram.filters import Command
from aiogram.fsm.context import FSMContext

from bot.config import CREDIT_PACKS, HOSTING_PLANS
from bot.database import get_user, get_user_bots, extend_hosting, update_credits

router = Router()


def credits_keyboard():
    buttons = []
    for key, pack in CREDIT_PACKS.items():
        buttons.append([
            InlineKeyboardButton(
                text=f"{pack['label']} — {pack['credits']} кред. | {pack['price_rub']} ₽ или {pack['stars']} ⭐",
                callback_data=f"buy_credits:{key}"
            )
        ])
    return InlineKeyboardMarkup(inline_keyboard=buttons)


def payment_method_keyboard(payload: str):
    return InlineKeyboardMarkup(inline_keyboard=[
        [InlineKeyboardButton(text="⭐ Telegram Stars", callback_data=f"pay_stars:{payload}")],
        [InlineKeyboardButton(text="💳 ЮMoney", callback_data=f"pay_yoomoney:{payload}")],
        [InlineKeyboardButton(text="❌ Отмена", callback_data="pay_cancel")],
    ])


def hosting_keyboard(bots: list):
    buttons = []
    for bot_item in bots:
        buttons.append([
            InlineKeyboardButton(
                text=f"🤖 {bot_item['name']}",
                callback_data=f"host_bot:{bot_item['id']}"
            )
        ])
    buttons.append([InlineKeyboardButton(text="❌ Отмена", callback_data="pay_cancel")])
    return InlineKeyboardMarkup(inline_keyboard=buttons)


def hosting_plans_keyboard(bot_id: int):
    buttons = []
    for key, plan in HOSTING_PLANS.items():
        buttons.append([
            InlineKeyboardButton(
                text=f"{plan['label']} — {plan['price_rub']} ₽ или {plan['stars']} ⭐",
                callback_data=f"buy_hosting:{key}:{bot_id}"
            )
        ])
    buttons.append([InlineKeyboardButton(text="◀️ Назад", callback_data="show_hosting")])
    return InlineKeyboardMarkup(inline_keyboard=buttons)


@router.message(Command("balance"))
@router.message(F.text == "💰 Баланс")
async def cmd_balance(message: Message):
    user = await get_user(message.from_user.id)
    if not user:
        await message.answer("Сначала отправь /start")
        return
    await message.answer(
        f"💰 Твой баланс: <b>{user['credits']} кредитов</b>\n\n"
        f"Кредиты используются для создания и улучшения ботов.\n"
        f"Хостинг оплачивается отдельно.",
        parse_mode="HTML"
    )


@router.message(Command("buy"))
@router.message(F.text == "🛒 Купить кредиты")
async def cmd_buy_credits(message: Message):
    await message.answer(
        "💳 Выбери пакет кредитов:\n\n"
        "🎁 <b>Новым пользователям</b> — 10 кредитов бесплатно при регистрации\n\n"
        "Кредиты нужны для:\n"
        "• Создать бота — 3 кредита\n"
        "• Добавить функцию — 2 кредита\n"
        "• Изменить логику — 1 кредит\n"
        "• Полная переработка — 3 кредита",
        parse_mode="HTML",
        reply_markup=credits_keyboard()
    )


@router.message(Command("hosting"))
@router.message(F.text == "🖥 Хостинг")
async def cmd_hosting(message: Message):
    user = await get_user(message.from_user.id)
    if not user:
        await message.answer("Сначала отправь /start")
        return

    bots = await get_user_bots(user["id"])
    if not bots:
        await message.answer("У тебя пока нет созданных ботов. Используй /create чтобы создать первого.")
        return

    lines = []
    for b in bots:
        until = b["hosting_until"] or "—"
        if until != "—":
            until = until[:10]
        status = "✅" if b["is_active"] else "⛔"
        lines.append(f"{status} <b>{b['name']}</b> — до {until}")

    await message.answer(
        "🖥 Твои боты:\n\n" + "\n".join(lines) + "\n\nВыбери бота для продления хостинга:",
        parse_mode="HTML",
        reply_markup=hosting_keyboard(bots)
    )


@router.callback_query(F.data == "show_hosting")
async def cb_show_hosting(callback: CallbackQuery):
    await callback.message.delete()
    await cmd_hosting(callback.message)
    await callback.answer()


@router.callback_query(F.data.startswith("host_bot:"))
async def cb_host_bot(callback: CallbackQuery):
    bot_id = int(callback.data.split(":")[1])
    await callback.message.edit_text(
        "Выбери период хостинга:",
        reply_markup=hosting_plans_keyboard(bot_id)
    )
    await callback.answer()


@router.callback_query(F.data.startswith("buy_credits:"))
async def cb_buy_credits(callback: CallbackQuery):
    pack_key = callback.data.split(":")[1]
    pack = CREDIT_PACKS[pack_key]
    await callback.message.edit_text(
        f"Пакет <b>{pack['label']}</b>: {pack['credits']} кредитов\n\n"
        f"Выбери способ оплаты:",
        parse_mode="HTML",
        reply_markup=payment_method_keyboard(f"credits:{pack_key}")
    )
    await callback.answer()


@router.callback_query(F.data.startswith("buy_hosting:"))
async def cb_buy_hosting(callback: CallbackQuery):
    _, plan_key, bot_id = callback.data.split(":")
    plan = HOSTING_PLANS[plan_key]
    await callback.message.edit_text(
        f"Хостинг <b>{plan['label']}</b>: {plan['price_rub']} ₽\n\n"
        f"Выбери способ оплаты:",
        parse_mode="HTML",
        reply_markup=payment_method_keyboard(f"hosting:{plan_key}:{bot_id}")
    )
    await callback.answer()


@router.callback_query(F.data.startswith("pay_stars:"))
async def cb_pay_stars(callback: CallbackQuery, bot: Bot):
    payload = callback.data[len("pay_stars:"):]
    parts = payload.split(":")

    if parts[0] == "credits":
        from bot.services.payments.telegram_stars import send_credits_invoice
        await send_credits_invoice(bot, callback.from_user.id, parts[1])
    elif parts[0] == "hosting":
        from bot.services.payments.telegram_stars import send_hosting_invoice
        await send_hosting_invoice(bot, callback.from_user.id, parts[1], int(parts[2]))

    await callback.message.delete()
    await callback.answer()


@router.callback_query(F.data.startswith("pay_yoomoney:"))
async def cb_pay_yoomoney(callback: CallbackQuery):
    payload = callback.data[len("pay_yoomoney:"):]
    parts = payload.split(":")
    user = await get_user(callback.from_user.id)

    if parts[0] == "credits":
        from bot.services.payments.yoomoney import make_credits_link
        link, label = make_credits_link(user["id"], parts[1])
        pack = CREDIT_PACKS[parts[1]]
        text = (
            f"💳 Оплата через ЮMoney\n\n"
            f"Пакет: {pack['label']} ({pack['credits']} кредитов)\n"
            f"Сумма: {pack['price_rub']} ₽\n\n"
            f"После оплаты кредиты начислятся автоматически.\n\n"
            f"👉 <a href='{link}'>Оплатить</a>"
        )
    elif parts[0] == "hosting":
        from bot.services.payments.yoomoney import make_hosting_link
        link, label = make_hosting_link(user["id"], parts[1], int(parts[2]))
        plan = HOSTING_PLANS[parts[1]]
        text = (
            f"💳 Оплата через ЮMoney\n\n"
            f"Хостинг: {plan['label']}\n"
            f"Сумма: {plan['price_rub']} ₽\n\n"
            f"После оплаты хостинг активируется автоматически.\n\n"
            f"👉 <a href='{link}'>Оплатить</a>"
        )
    else:
        await callback.answer("Ошибка")
        return

    await callback.message.edit_text(text, parse_mode="HTML", disable_web_page_preview=True)
    await callback.answer()


@router.callback_query(F.data == "pay_cancel")
async def cb_pay_cancel(callback: CallbackQuery):
    await callback.message.delete()
    await callback.answer("Отменено")


@router.pre_checkout_query()
async def pre_checkout(query: PreCheckoutQuery):
    await query.answer(ok=True)


@router.message(F.successful_payment)
async def successful_payment(message: Message):
    payload = message.successful_payment.invoice_payload
    parts = payload.split(":")
    user = await get_user(message.from_user.id)

    if parts[0] == "credits":
        from bot.config import CREDIT_PACKS
        pack = CREDIT_PACKS[parts[1]]
        await update_credits(
            user["id"], pack["credits"],
            description=f"Покупка пакета {pack['label']}",
            payment_method="telegram_stars",
            amount_rub=0
        )
        await message.answer(
            f"✅ Оплата прошла! Начислено <b>{pack['credits']} кредитов</b>.\n\n"
            f"Используй /create чтобы создать нового бота.",
            parse_mode="HTML"
        )

    elif parts[0] == "hosting":
        from bot.config import HOSTING_PLANS
        plan = HOSTING_PLANS[parts[1]]
        bot_id = int(parts[2])
        new_until = await extend_hosting(bot_id, plan["days"])
        await message.answer(
            f"✅ Хостинг оплачен! Бот будет работать до <b>{new_until[:10]}</b>.",
            parse_mode="HTML"
        )
